python/cudf/bindings/csv.pyx

```diff
@@ -0,0 +1,452 @@
+    ... ]
+    >>> with open(filename, 'w') as fp:
+    ...     fp.write('\\n'.join(lines)+'\\n')
+
+    Read the file with ``cudf.read_csv``
+
+    >>> cudf.read_csv(filename)
+      num1                datetime text
+    0  123 2018-11-13T12:00:00.000 5451
+    1  456 2018-11-14T12:35:01.000 5784
+    2  789 2018-11-15T18:02:59.000 6117
+    """
+
+    if delim_whitespace:
+        if delimiter is not None:
+            raise ValueError("cannot set both delimiter and delim_whitespace")
+        if sep != ',':
+            raise ValueError("cannot set both sep and delim_whitespace")
+
+    # Alias sep -> delimiter.
+    if delimiter is None:
+        delimiter = sep
+
+    if dtype is not None:
+        if isinstance(dtype, collections.abc.Mapping):
+            dtype_dict = True
+        elif isinstance(dtype, collections.abc.Iterable):
+            dtype_dict = False
+        else:
+            msg = '''dtype must be 'list like' or 'dict' '''
+            raise TypeError(msg)
+        if names is not None and len(dtype) != len(names):
+            msg = '''All column dtypes must be specified.'''
+            raise TypeError(msg)
+
+    nvtx_range_push("CUDF_READ_CSV", "purple")
+
+    # csv_reader = ffi.new('csv_read_arg*')
+    cdef csv_read_arg csv_reader = csv_read_arg()
+
+    # Populate csv_reader struct
+    cdef char* buffer_data_holder
+    if is_file_like(filepath_or_buffer):
+        if compression == 'infer':
+            compression = None
+        buffer = filepath_or_buffer.read()
+        # check if StringIO is used
+        if hasattr(buffer, 'encode'):
+            buffer_as_bytes = buffer.encode()
+        else:
+            buffer_as_bytes = buffer
+        # buffer_data_holder = ffi.new("char[]", buffer_as_bytes)
+        buffer_data_holder = buffer_as_bytes
+
+        csv_reader.input_data_form = HOST_BUFFER
+        csv_reader.filepath_or_buffer = buffer_data_holder
+        csv_reader.buffer_size = len(buffer_as_bytes)
+    else:
+        if (not os.path.isfile(filepath_or_buffer)):
+            raise(FileNotFoundError)
+        if (not os.path.exists(filepath_or_buffer)):
+            raise(FileNotFoundError)
+        file_path = filepath_or_buffer.encode()
+
+        csv_reader.input_data_form = FILE_PATH
+        csv_reader.filepath_or_buffer = file_path
+
+    if header == 'infer':
+        header = -1
+    header_infer = header
+    arr_names = []
+    cdef vector[const char*] vector_names
+    arr_dtypes = []
+    cdef vector[const char*] vector_dtypes
+    if names is None:
+        if header is -1:
+            header_infer = 0
+        if header is None:
+            header_infer = -1
+        # csv_reader.names = ffi.NULL
+        csv_reader.names = NULL
+        csv_reader.num_cols = 0
+        if dtype is not None:
+            if dtype_dict:
+                for idx, (k, v) in enumerate(dtype.items()):
+                    arr_dtypes.append(_wrap_string(str(str(k)+":"+str(v))))
```

Comment:
I believe this only works as expected because the `<char*> output` will get converted to Python bytes automagically.